File: src/services/public-url.ts
import { type } from "arktype";

/**
 * Returns true when `value` is a syntactically valid `http(s)` URL whose host
 * is expected to be resolvable on the public internet.
 *
 * The following host classes are rejected:
 *  - `localhost` and any `*.localhost` / `*.local` hostnames
 *  - IPv4 loopback `127.0.0.0/8` and the unspecified address `0.0.0.0`
 *  - IPv6 loopback `::1` and unspecified `::`
 *  - IPv4 link-local `169.254.0.0/16`
 *  - IPv6 link-local `fe80::/10`
 *  - RFC1918 private IPv4 ranges: `10.0.0.0/8`, `172.16.0.0/12`,
 *    `192.168.0.0/16`
 *
 * This is enforced client-side so users get an actionable error immediately
 * rather than an opaque 403 from the upstream WAF.
 */
export function isPublicRoutableUrl(value: string): boolean {
	if (typeof value !== "string" || value.trim() === "") {
		return false;
	}

	let parsed: URL;
	try {
		parsed = new URL(value);
	} catch {
		return false;
	}

	if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
		return false;
	}

	const rawHost = parsed.hostname.toLowerCase();

	if (rawHost === "") {
		return false;
	}

	// WHATWG URL returns IPv6 hostnames wrapped in square brackets.
	const host =
		rawHost.startsWith("[") && rawHost.endsWith("]")
			? rawHost.slice(1, -1)
			: rawHost;

	if (host === "localhost") {
		return false;
	}

	if (host.endsWith(".localhost") || host.endsWith(".local")) {
		return false;
	}

	if (isIPv4(host)) {
		return isPublicIPv4(host);
	}

	if (isIPv6(host)) {
		return isPublicIPv6(host);
	}

	return true;
}

function isIPv4(host: string): boolean {
	return /^(\d{1,3}\.){3}\d{1,3}$/.test(host);
}

function parseIPv4Octets(host: string): number[] | null {
	const parts = host.split(".").map((p) => Number(p));
	if (parts.length !== 4) return null;
	for (const p of parts) {
		if (!Number.isInteger(p) || p < 0 || p > 255) return null;
	}
	return parts;
}

function isPublicIPv4(host: string): boolean {
	const octets = parseIPv4Octets(host);
	if (!octets) return false;
	const [a, b] = octets as [number, number, number, number];

	// Unspecified 0.0.0.0/8
	if (a === 0) return false;
	// Loopback 127.0.0.0/8
	if (a === 127) return false;
	// RFC1918: 10.0.0.0/8
	if (a === 10) return false;
	// RFC1918: 172.16.0.0/12
	if (a === 172 && b >= 16 && b <= 31) return false;
	// RFC1918: 192.168.0.0/16
	if (a === 192 && b === 168) return false;
	// Link-local 169.254.0.0/16
	if (a === 169 && b === 254) return false;

	return true;
}

function isIPv6(host: string): boolean {
	// WHATWG URL parses IPv6 hostnames wrapped in [] and returns them
	// unwrapped in .hostname, so detect by the presence of a ":".
	return host.includes(":");
}

function isPublicIPv6(host: string): boolean {
	const normalized = host.toLowerCase();

	if (normalized === "::" || normalized === "::1") return false;
	if (normalized.startsWith("fe80:") || normalized.startsWith("fe80::")) {
		return false;
	}

	return true;
}

/**
 * Arktype narrowing that accepts only publicly-routable `http(s)` URLs.
 *
 * Use in schemas in place of `type.keywords.string.url.root` when the URL
 * will be stored server-side and needs to be reachable from the public
 * internet (for example: app homepage URLs and OAuth redirect URIs).
 */
export const publicHttpUrl = type("string").narrow((value, ctx) => {
	if (isPublicRoutableUrl(value)) {
		return true;
	}
	return ctx.mustBe(
		"a publicly-resolvable http(s) URL (localhost, loopback, and private IPs are not allowed)",
	);
});

File: rust/src/application/public_url.rs
use url::{Host, Url};

/// Returns true when `value` is a syntactically valid `http(s)` URL whose host
/// is expected to be resolvable on the public internet.
///
/// The following host classes are rejected:
///  - `localhost` and any `*.localhost` / `*.local` hostnames
///  - IPv4 loopback `127.0.0.0/8`, unspecified `0.0.0.0/8`
///  - IPv6 loopback `::1` and unspecified `::`
///  - IPv4 link-local `169.254.0.0/16`
///  - IPv6 link-local `fe80::/10`
///  - RFC1918 private IPv4 ranges: `10.0.0.0/8`, `172.16.0.0/12`, `192.168.0.0/16`
///  - CGNAT / shared address space `100.64.0.0/10` (RFC 6598)
///  - IPv6 unique local addresses `fc00::/7` (RFC 4193)
///  - IPv6 reserved `fe00::/9` (not globally routable)
///
/// Enforced client-side so users get an actionable error immediately rather than
/// an opaque 403 from the upstream WAF.
pub fn is_public_routable_url(value: &str) -> bool {
    let Ok(url) = Url::parse(value) else {
        return false;
    };

    if url.scheme() != "http" && url.scheme() != "https" {
        return false;
    }

    match url.host() {
        Some(Host::Domain(domain)) => {
            let host = domain.to_ascii_lowercase();
            !(host == "localhost"
                || host == "local"
                || host.ends_with(".localhost")
                || host.ends_with(".local"))
        }
        Some(Host::Ipv4(addr)) => {
            let octets = addr.octets();
            let is_shared = octets[0] == 100 && (64..=127).contains(&octets[1]);
            !(octets[0] == 0
                || addr.is_loopback()
                || addr.is_private()
                || addr.is_link_local()
                || is_shared)
        }
        Some(Host::Ipv6(addr)) => {
            let seg0 = addr.segments()[0];
            let is_unique_local = (seg0 & 0xfe00) == 0xfc00; // fc00::/7
            let is_reserved_fe00 = (seg0 & 0xff80) == 0xfe00; // fe00::/9
            !(addr.is_loopback()
                || addr.is_unspecified()
                || addr.is_unicast_link_local()
                || is_unique_local
                || is_reserved_fe00)
        }
        None => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn accepts_public_https_url() {
        assert!(is_public_routable_url("https://example.com"));
        assert!(is_public_routable_url("https://[2606:4700::1111]"));
    }

    #[test]
    fn rejects_non_http_scheme() {
        assert!(!is_public_routable_url("ftp://example.com/x"));
    }

    #[test]
    fn rejects_localhost_and_dot_local() {
        assert!(!is_public_routable_url("http://localhost"));
        assert!(!is_public_routable_url("http://local"));
        assert!(!is_public_routable_url("https://api.localhost"));
        assert!(!is_public_routable_url("https://api.local"));
    }

    #[test]
    fn rejects_loopback_and_private_ipv4() {
        assert!(!is_public_routable_url("http://127.0.0.1"));
        assert!(!is_public_routable_url("http://0.0.0.0"));
        assert!(!is_public_routable_url("http://10.0.0.1"));
        assert!(!is_public_routable_url("http://172.16.0.1"));
        assert!(!is_public_routable_url("http://192.168.1.1"));
        assert!(!is_public_routable_url("http://169.254.1.1"));
    }

    #[test]
    fn rejects_cgnat_shared_address_space() {
        assert!(!is_public_routable_url("http://100.64.0.1"));
        assert!(!is_public_routable_url("http://100.127.255.255"));
        assert!(is_public_routable_url("http://100.63.255.255"));
        assert!(is_public_routable_url("http://100.128.0.1"));
    }

    #[test]
    fn rejects_loopback_and_link_local_ipv6() {
        assert!(!is_public_routable_url("https://[::]"));
        assert!(!is_public_routable_url("https://[::1]"));
        assert!(!is_public_routable_url("https://[fe80::1]"));
    }

    #[test]
    fn rejects_unique_local_ipv6() {
        assert!(!is_public_routable_url("https://[fc00::1]"));
        assert!(!is_public_routable_url("https://[fd12:3456:789a::1]"));
    }

    #[test]
    fn rejects_reserved_fe00_ipv6() {
        assert!(!is_public_routable_url("https://[fe00::1]"));
        assert!(!is_public_routable_url("https://[fe7f::1]"));
    }
}
